File: clojure/03-operators/arithmetic.clj
;; Basic arithmetic - all functions are variadic
(println "Addition:        " (+ 2 3 4))
(println "Subtraction:     " (- 20 5 3))
(println "Multiplication:  " (* 2 3 4))
(println "Division:        " (/ 20 4))

;; Integer vs ratio division
(println "Ratio result:    " (/ 10 3))      ; => 10/3 (exact ratio)
(println "Float division:  " (/ 10.0 3))    ; => 3.333...
(println "Integer quotient:" (quot 10 3))   ; => 3
(println "Remainder:       " (rem 10 3))    ; => 1
(println "Modulo:          " (mod -7 3))    ; => 2 (mod differs from rem for negatives)

;; Exponentiation and absolute value
(println "Power 2^10:      " (Math/pow 2 10))
(println "Absolute value:  " (abs -42))

;; Increment and decrement helpers
(println "Increment:       " (inc 5))
(println "Decrement:       " (dec 5))

;; Min and max are also variadic
(println "Min of values:   " (min 3 1 4 1 5 9 2 6))
(println "Max of values:   " (max 3 1 4 1 5 9 2 6))

File: lua/03-operators/arithmetic.lua
-- Basic arithmetic
local a = 17
local b = 5

print("a + b  =", a + b)    -- Addition
print("a - b  =", a - b)    -- Subtraction
print("a * b  =", a * b)    -- Multiplication
print("a / b  =", a / b)    -- Float division (always returns float)
print("a // b =", a // b)   -- Floor division (since Lua 5.3)
print("a % b  =", a % b)    -- Modulo
print("a ^ b  =", a ^ b)    -- Exponentiation (always returns float)
print("-a     =", -a)       -- Unary minus

-- Integer vs float distinction in Lua 5.3+
print("Type of 10 / 2  :", math.type(10 / 2))    -- float (/ is float division)
print("Type of 10 // 2 :", math.type(10 // 2))   -- integer
print("Type of 2 ^ 3   :", math.type(2 ^ 3))     -- float
